pkg/statistics/histogram.go

```diff
@@ -593,14 +593,16 @@ func (hg *Histogram) LessRowCount(sctx context.PlanContext, value types.Datum) f
 func (hg *Histogram) BetweenRowCount(sctx context.PlanContext, a, b types.Datum) float64 {
 	lessCountA := hg.LessRowCount(sctx, a)
 	lessCountB := hg.LessRowCount(sctx, b)
-	// If lessCountA is not less than lessCountB, it may be that they fall to the same bucket and we cannot estimate
-	// the fraction, so we use `totalCount / NDV` to estimate the row count, but the result should not greater than
-	// lessCountB or notNullCount-lessCountA.
-	if lessCountA >= lessCountB && hg.NDV > 0 {
+	rangeEst := lessCountB - lessCountA
+	aEqual, _ := hg.EqualRowCount(sctx, a, false)
+	bEqual, _ := hg.EqualRowCount(sctx, b, false)
```

Comment:
Before we enter the `BetweenRowCount`, we'll modify the `Range` struct to ensure that the passed range here is always in the form `[a, x)`.

If the given filter is `a >= 1 and a <= 3` => the range is [1, 3], the range passed to the `BetweenRowCount` will be something like `[1, 4)`.

So directly using `EqualRowCount(a)` and `EqualRowCount(b)` may also lead to wrong estimation.

So my suggestion is that:
- If the Range.Lower.Exclude is false and the `Lower` itself is the upper bound of one bucket, use `Bucket.Repeat` which is the occurrence of the value.
- For any other cases for the `Range.Lower`, use `notNullCount()/NDV`.
- For `Range.Upper`, it's really hard to do the equal row count estimation since it's always excluded from the range we do the estimation. I actually have no good idea about this part, just suggest using the `notNullCount()/NDV`.

Reply:
If range lower is false - the call for equals estimation will already return notNullCount()/NDV. For Range.Upper - as per the review comment here - I have replaced it with notNullCount()/NDV.